### PR TITLE
Dockerize

### PR DESCRIPTION
**Pull request by nick-pestell**
**This PR implements refactoring to the python code and a dockerized workflow for using with ViewBovine as well as adding a `Readme.md`**

### Refactors:
Most of the python code remains unchanged, however the original `btb_phylo.py` has now been split into two separate modules, `phylogeny.py` and `filter_samples.py`. The interface code has all been moved into a new python file called `btb_phylo.py`.
The code now consists of three main modules:
1. `update_summary.py` - scans s3 for new batches and updates a local copy of the summary `.csv`
2. `filter_samples.py` - filters samples for phylogeny.
3.  `phylogeny.py` - downloads consensus sequences, builds giant fasta  file calls `snp-sites`, `snp-dists` and optionally `megacc`.
As, mentioned, a fourth file `btb_phylo.py` serves as an interface which can run the entire pipeline or sections of it in isolation via command line arguments (sub-commands).

### Dockerized workflow:
- A bash script, `btb-phylo`, will run the full pipeline when called.
- Required positional arguments include paths to the results folder and consensus folder (where consensus sequences are downloaded to) 
- The script contains a function `parse_args()` which parses optional arguments:
1. `-c` for a configuration file, (json file format describing filtering parameters, [see example](https://github.com/APHA-CSU/btb-phylo/blob/dockerize/example_config.json)) 
2. `-j` for the number of threads to use with `snp-dists` 
3. `with-docker`.  
- If `with-docker` flag is used when calling `btb-pyhlo` then the latest docker image is pulled from DockerHub and subsequently run, with bind mounts to the results directory, consensus directory and config file. These paths are also passed in as arguments to the `docker run` command.
- The entry point to the docker container (see `Dockerfile` line 28) is `btb-phylo`, which is called subsequently called **without** the `with-docker` flag and so passes to line 81, calling `python btb_phylo.py full_pipeline`.

## `Readme.md`
This describes how to install and run the software with a number of different configurations and how it should be used to serve ViewBovine APHA.

### CODE REVIEW

**Review by CameroneN — APPROVED** [review]
Looks good, I dont fully understand docker hahah but the new readme is great and I think the refactoring of phylo is good

File: btb-phylo [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    if (( $#<2 ))
+    then
+        printf "Must include positional arguments of paths to: results folder "
+        printf "and consensus folder\n"
+        exit 1
+    fi
+
+    RESULTS=$1
+    CONSENSUS=$2
+    shift
```
Comment (ellisrichardj) [inline btb-phylo:46]: Could replace with `shift shift` with `shift 2`

File: btb-phylo [hunk 2]
```diff
@@ -0,0 +1,82 @@
+#!/bin/bash
```
Comment (ellisrichardj) [inline btb-phylo:1]: Suggest renaming as `btb-phylo.sh` to indicate that this is a bash script

**Review by ellisrichardj — COMMENTED** [review]
Nice work to split out the various components of the processing.  Just a few comments on the bash script

File: btb-phylo (hunk 2 above)
Comment (nick-pestell) [reply btb-phylo:1]: Cool, done

File: btb-phylo [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        else
+        printf "Unrecognised argument: '%s'. Please include positional arguments for " "$1"
+        printf "paths to: results folder and consensus folder. Additional optional arguments "
+        printf "include 'with-docker', '-j' and '-c' (see Readme)\n"
+        exit 1
+        fi
+        shift
+    done
+}
+
```
[the remark this answers is not in the document]
Comment (nick-pestell) [reply btb-phylo]: Yeah, you may well be right. This was my best attempt at having named optional arguments. The issue I was running into with using standard defaults was that the order made a difference. E.g. I couldn't work out how to  omit arg 2 (go with default) but use arg 3. 

Ref: https://stackoverflow.com/questions/71349910/optional-arguments-in-bash-script

**Pull request merged**